From e5bd1f8e0a086e3502cbf928dbbe2be7f8d8b59a Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Sat, 28 Jul 2018 00:49:05 +0800
Subject: [PATCH 4250/5242] MGS-4103 [#imx-1070] fix gpu axi bus error on 6sx

commit  886ebe52bb5db1ab4ddcc85fca5f115172a662eb from
https://source.codeaurora.org/external/imx/linux-imx.git

only add the padding pages for user pageable memory,

Revert "MGS-4022 [#imx-1070] fix kernel panic with opencl test_buffers"
This reverts commit 9253786bcfd2fa17132c5057a18eb75b10ea3336.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../default/gc_hal_kernel_allocator_user_memory.c  |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
index 8258cf8..301db84 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
@@ -482,6 +482,12 @@ static int import_pfn_map(struct um_desc *um,
     UserMemory->pageCount = pageCount;
     UserMemory->extraPage = extraPage;
 
+    if (extraPage && UserMemory->type == UM_PAGE_MAP)
+    {
+        /*Add the padding pages */
+        UserMemory->chunk_count++;
+    }
+
     /* Success. */
     gcmkFOOTER();
     return gcvSTATUS_OK;
@@ -516,8 +522,7 @@ static int import_pfn_map(struct um_desc *um,
 
     Mdl->priv = userMemory;
     Mdl->numPages = userMemory->pageCount + userMemory->extraPage;
-    /* Not contiguous with extra pages */
-    Mdl->contiguous = gcvFALSE;
+    Mdl->contiguous = (userMemory->chunk_count == 1);
 
     gcmkFOOTER_NO();
     return gcvSTATUS_OK;
-- 
1.7.9.5

